File: backend/app/api/routes/simplex.py
from fastapi import APIRouter, HTTPException
from app.models.schemas import SimplexRequest, SimplexResponse, TableauRow, IterationTableau
from app.core.simplex_engine import SimplexEngine
from app.core.graphical_method import build_graphical_data

router = APIRouter(prefix="/simplex", tags=["simplex"])


@router.post("/solve", response_model=SimplexResponse, summary="Solve a linear programme via the Simplex method")
def solve(payload: SimplexRequest) -> SimplexResponse:
    """
    Accepts an objective function (2-4 variables), an optimisation goal
    (max/min) and a list of linear constraints, then returns the optimal
    solution together with the final simplex tableau.
    """
    try:
        engine = SimplexEngine(
            objective=payload.objective,
            goal=payload.goal,
            constraint_coeffs=[c.coefficients for c in payload.constraints],
            inequalities=[c.inequality for c in payload.constraints],
            rhs_values=[c.rhs for c in payload.constraints],
        )
        result = engine.solve()
    except Exception as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc

    graphical = None
    if len(payload.objective) == 2:
        optimal_point = None
        if result.status == "optimal" and result.variables is not None:
            optimal_point = {
                "x": result.variables.get("x1", 0.0),
                "y": result.variables.get("x2", 0.0),
            }
        graphical = build_graphical_data(
            objective=payload.objective,
            goal=payload.goal,
            constraint_coeffs=[c.coefficients for c in payload.constraints],
            inequalities=[c.inequality for c in payload.constraints],
            rhs_values=[c.rhs for c in payload.constraints],
            status=result.status,
            optimal_point=optimal_point,
            objective_value=result.objective_value,
        )

    def _build_iteration_tableaux(snapshots: list[dict]) -> list[IterationTableau]:
        return [
            IterationTableau(
                iteration=s["iteration"],
                entering=s.get("entering"),
                leaving=s.get("leaving"),
                tableau_headers=s["tableau_headers"],
                tableau_rows=[
                    TableauRow(basic_variable=r["basic_variable"], values=r["values"])
                    for r in s["tableau_rows"]
                ],
            )
            for s in snapshots
        ]

    if result.status == "optimal":
        return SimplexResponse(
            status="optimal",
            objective_value=result.objective_value,
            variables=result.variables,
            iterations=result.iterations,
            tableau_headers=result.tableau_headers,
            tableau_rows=[
                TableauRow(basic_variable=r["basic_variable"], values=r["values"])
                for r in result.tableau_rows
            ],
            message=result.message,
            graphical=graphical,
            iteration_tableaux=_build_iteration_tableaux(result.iteration_tableaux),
        )

    return SimplexResponse(
        status=result.status,
        message=result.message,
        graphical=graphical,
    )


@router.get("/health", summary="Health check")
def health() -> dict:
    return {"status": "ok", "service": "simplex-optimizer-api"}

File: frontend/src/types/simplex.ts
// ─── Domain types ────────────────────────────────────────────────────────────

export type Goal = "max" | "min";
export type Inequality = "<=" | ">=" | "=";
export type VariableCount = 2 | 3 | 4;
export type SolveStatus = "optimal" | "unbounded" | "infeasible";

export interface Constraint {
  id: string;
  coefficients: number[];
  inequality: Inequality;
  rhs: number;
}

export interface ObjectiveFunction {
  coefficients: number[];
  goal: Goal;
}

// ─── API request / response ───────────────────────────────────────────────────

export interface SimplexRequest {
  objective: number[];
  goal: Goal;
  constraints: Array<{
    coefficients: number[];
    inequality: Inequality;
    rhs: number;
  }>;
}

export interface TableauRow {
  basic_variable: string;
  values: number[];
}

export interface GraphPoint {
  x: number;
  y: number;
}

export interface GraphLine {
  a: number;
  b: number;
  c: number;
  points: GraphPoint[];
  label?: string;
  inequality?: Inequality;
}

export interface GraphBounds {
  x_min: number;
  x_max: number;
  y_min: number;
  y_max: number;
}

export interface GraphicalData {
  status: SolveStatus;
  bounds: GraphBounds;
  constraints: GraphLine[];
  feasible_polygon?: GraphPoint[] | null;
  vertices: GraphPoint[];
  objective_line?: GraphLine | null;
  optimal_point?: GraphPoint | null;
  objective_value?: number | null;
  goal: Goal;
}

export interface IterationTableau {
  iteration: number;
  entering?: string | null;
  leaving?: string | null;
  tableau_headers: string[];
  tableau_rows: TableauRow[];
}

export interface SimplexResponse {
  status: SolveStatus;
  objective_value?: number;
  variables?: Record<string, number>;
  iterations?: number;
  tableau_headers?: string[];
  tableau_rows?: TableauRow[];
  message?: string;
  graphical?: GraphicalData | null;
  iteration_tableaux?: IterationTableau[] | null;
}

// ─── UI state ─────────────────────────────────────────────────────────────────

export interface SolverState {
  nVars: VariableCount;
  objective: ObjectiveFunction;
  constraints: Constraint[];
}
